rskj-core/src/test/java/co/rsk/peg/federation/FederationStorageProviderImplTests.java

```diff
@@ -979,6 +981,110 @@ void getNextFederationCreationBlockHeight_afterRSKIP186AndNoValueInStorage_retur
+        byte[] serializedScript = BridgeSerializationUtils.serializeScript(expectedScript);
+        // Putting some value in the storage just to then assert that before fork, the storage won't be accessed.
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), serializedScript);
+
+        // Act
+
+        FederationStorageProvider federationStorageProvider = new FederationStorageProviderImpl(storageAccessor);
+
+        // Assert
+        assertEquals(Optional.empty(), federationStorageProvider.getLastRetiredFederationP2SHScript(activations));
+
+    }
+
+    @Test
+    void getLastRetiredFederationP2SHScript_afterRSKIP186_getsValueFromStorage() {
+
+        // Arrange
+
+        ActivationConfig.ForBlock activations = mock(ActivationConfig.ForBlock.class);
+        when(activations.isActive(ConsensusRule.RSKIP186)).thenReturn(true);
+
+        StorageAccessor storageAccessor = new InMemoryStorage();
+        Script expectedScript = new P2shErpFederationBuilder().build().getP2SHScript();
+
+        byte[] serializedScript = BridgeSerializationUtils.serializeScript(expectedScript);
+
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), serializedScript);
+
+        // Act
+
+        FederationStorageProvider federationStorageProvider = new FederationStorageProviderImpl(storageAccessor);
+        Optional<Script> actualValueOptional = federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
+
+        // Assert
+
+        assertTrue(actualValueOptional.isPresent());
+        assertEquals(expectedScript, actualValueOptional.get());
+
+        // Setting in storage a different value to assert that calling the method again should return cached value
+
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), expectedScript.getProgram());
+
+        Optional<Script> actualCachedValue = federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
+
+        assertTrue(actualCachedValue.isPresent());
+        assertEquals(expectedScript, actualCachedValue.get());
+
+    }
+
+    @Test
+    void getLastRetiredFederationP2SHScript_afterRSKIP186AndNoValueInStorage_returnsEmpty() {
+
+        // Arrange
+
+        ActivationConfig.ForBlock activations = mock(ActivationConfig.ForBlock.class);
+        when(activations.isActive(ConsensusRule.RSKIP186)).thenReturn(true);
+
+        StorageAccessor storageAccessor = new InMemoryStorage();
+        storageAccessor.saveToRepository(LAST_RETIRED_FEDERATION_P2SH_SCRIPT_KEY.getKey(), null);
+
+        // Act
+
+        FederationStorageProvider federationStorageProvider = new FederationStorageProviderImpl(storageAccessor);
+        Optional<Script> actualValueOptional = federationStorageProvider.getLastRetiredFederationP2SHScript(activations);
+
+        // Assert
+
+        assertFalse(actualValueOptional.isPresent());
```

Comment:
we should define what we wanna assert, if this or [this](https://github.com/rsksmart/rskj/pull/2547/files#diff-f42b277f39730e846e15f899b88e1e7e718280d3ea3d4c4d5b035a5d66385923R1004)

Reply:
This case is different. If I do `assertEquals(Optional.empty(), actualValueOptional.get());` the compiler will throw a warning because I'm calling `Optional::get` without checking checking first `Optional::isPresent`. To avoid that warning, I need to call `Optional::isPresent` first. If I do, then there's no need to call `Optional::get` if there's no value present.